helix-rest/src/main/java/org/apache/helix/rest/metadatastore/accessor/ZkRoutingDataWriter.java

```diff
@@ -0,0 +1,265 @@
+package org.apache.helix.rest.metadatastore.accessor;
+
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+
+import org.I0Itec.zkclient.exception.ZkNodeExistsException;
+import org.apache.helix.ZNRecord;
+import org.apache.helix.manager.zk.ZNRecordSerializer;
+import org.apache.helix.manager.zk.ZkBaseDataAccessor;
+import org.apache.helix.manager.zk.client.DedicatedZkClientFactory;
+import org.apache.helix.manager.zk.client.HelixZkClient;
+import org.apache.helix.rest.metadatastore.concurrency.DistributedLock;
+import org.apache.helix.rest.metadatastore.concurrency.ZkDistributedLock;
+import org.apache.helix.rest.metadatastore.constant.MetadataStoreRoutingConstants;
+import org.apache.helix.rest.metadatastore.exceptions.ZkLockException;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+public class ZkRoutingDataWriter implements MetadataStoreRoutingDataWriter {
+  private static final Logger LOG = LoggerFactory.getLogger(ZkBaseDataAccessor.class);
+
+  private final String _namespace;
+  private final String _zkAddress;
+  private final HelixZkClient _zkClient;
+  private final DistributedLock _routingDataLock;
```

Comment:
Fundamentally, I think this is not an efficient way for purpose of control of centralized work. This may increase the traffic to ZK. 

It will be better to work as Helix leader controller. One of the REST server become write REST server (acquire the lock once). Write request to read-only server forward the request to write server for write purpose.

Reply:
@dasahcc 
I agree that making Helix API service (Helix REST) an Helix application is a clever solution, but I have some concerns about that approach - wouldn't we be creating a circular dependency?

Also, there was an agreement on this approach because we have an assumption that write operations are rare. Ideally, all operations provided by the writer would be ACL-ed and will only be done by admin. In that sense, I believe this is an appropriate approach since it is simple, self-contained without any other dependency (say Helix controller), and it follows a distributed lock recipe that ZooKeeper is very well-known and tested for.

Reply:
Update: using distributed leader election.